Add option to change resolve path for image

By default, image path when calculating image size is resolved by using `process.cwd()` as baseline (https://github.com/simonsmith/postcss-at2x/blob/master/lib/index.js#L91). This is not always the case (maybe never?).

It would be nice to have option to change that and maybe by default use input CSS `from` property (`decl.source.input.from`) or use `process.cwd()` by default and provide source information as callback argument for developers.

E.g. 

```js
resolveImagePath: ( source ) => {
	return process.cwd();
	// `source` contains input CSS information
}
```

Hi @simonsmith, sorry for waiting this long to work on this issue.

I’ve created PR for this, but I’m not able to write tests, I’m getting following error from Jest:

```sh
> postcss-at2x@3.0.0 test /Users/niksy/Projects/postcss-at2x
> jest && npm run lint

 FAIL  test/test.js
  ● Test suite failed to run

    SyntaxError: Unexpected reserved word
      
      at transformAndBuildScript (node_modules/jest/node_modules/jest-cli/node_modules/jest-runtime/build/transform.js:320:12)
      at process._tickCallback (node.js:364:9)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        1.276s
Ran all test suites.
npm ERR! Test failed.  See above for more details.
```

I’ve tried to work around it, but I was unable to, event after following several issues which explain the exact problem. Do you maybe know what’s wrong?

--- a/README.md
+++ b/README.md
@@ -77,6 +77,12 @@ Change the identifier added to retina images, for example `file@2x.png` can be `
 Obtains the image dimensions of the non-retina image automatically and applies them to the
 `background-size` property of the retina image.
 
+##### `resolveImagePath` _function_
+
+Get resolved image path for detecting image size. By default, original `url` value is resolved from current working directory (`process.cwd()`).
+
+Function receives two arguments: original `url` value and [PostCSS declaration source](http://api.postcss.org/Declaration.html#source).
+
 **Output**
 
 ```css

--- a/lib/index.js
+++ b/lib/index.js
@@ -12,9 +12,13 @@ const defaultResolutions = [
   '(min-resolution: 1.5dppx)',
 ];
 
+function defaultResolveImagePath (value) {
+  return path.resolve(process.cwd(), value);
+}
+
 export default postcss.plugin('postcss-at2x', at2x);
 
-function at2x({identifier = '@2x', detectImageSize = false} = {}) {
+function at2x({identifier = '@2x', detectImageSize = false, resolveImagePath = defaultResolveImagePath} = {}) {
   return function(root) {
     // Create an empty rule so that all the new rules can be appended to this
     // and then append it at the end.
@@ -48,7 +52,12 @@ function at2x({identifier = '@2x', detectImageSize = false} = {}) {
         // needs it for regex search
         decl.value = removeKeyword(decl.value);
 
-        const promise = getBackgroundImageSize(decl, backgroundSize, detectImageSize)
+        const promise = getBackgroundImageSize(
+          decl,
+          backgroundSize,
+          detectImageSize,
+          resolveImagePath
+        )
           .then((size) => {
             if (size) {
               retinaRule.append(postcss.decl(size));
@@ -74,7 +83,7 @@ function at2x({identifier = '@2x', detectImageSize = false} = {}) {
   };
 }
 
-function getBackgroundImageSize(decl, existingBackgroundSize, detectImageSize) {
+function getBackgroundImageSize(decl, existingBackgroundSize, detectImageSize, resolveImagePath) {
   if (!detectImageSize) {
     return Promise.resolve(existingBackgroundSize);
   }
@@ -88,7 +97,7 @@ function getBackgroundImageSize(decl, existingBackgroundSize, detectImageSize) {
     }
     node.nodes.forEach((fp) => {
       if (!path.isAbsolute(fp.value) && !isUrl(fp.value)) {
-        urlValue = path.resolve(process.cwd(), fp.value);
+        urlValue = resolveImagePath(fp.value, decl.source);
       }
     });
   });

--- a/test/fixtures/resolve-image-path.css
+++ b/test/fixtures/resolve-image-path.css
@@ -1,0 +1,3 @@
+.cat {
+  background-image: url('./test/fixtures/images/cat.jpg') at-2x;
+}

--- a/test/test.js
+++ b/test/test.js
@@ -12,7 +12,7 @@ function runTest(input, opts, done) {
 
   postcss([plugin(opts)]).process(input).then((result) => {
     expect(result.css).toMatchSnapshot();
-    expect(result.warnings()).toBeEmpty();
+    expect(result.warnings()).toHaveLength(0);
     done();
   }).catch(done);
 }
@@ -36,5 +36,23 @@ describe('postcss-at2x', () => {
 
   it('should process image and add background size', (done) => {
     runTest('read-background-size.css', {detectImageSize: true}, done);
+  });
+
+  it('should resolve image path with custom function', (done) => {
+    runTest('resolve-image-path.css', {
+      detectImageSize: true,
+      resolveImagePath: (value, source) => {
+        expect(typeof value).toBe('string');
+        expect(source).toHaveProperty('input.css');
+        return path.resolve(process.cwd(), value);
+      }}, done);
+  });
+
+  it('should have proper arguments for resolve image path function', (done) => {
+    const spy = jest.fn();
+    runTest('resolve-image-path.css', {detectImageSize: true, resolveImagePath: spy}, done);
+    const spyCall = spy.mock.calls[0];
+    expect(spyCall[0]).toEqual('./test/fixtures/images/cat.jpg');
+    expect(spyCall[1]).toMatchSnapshot();
   });
 });

